lib/parallel/parallel.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package parallel
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os/exec"
+	"runtime"
+	"time"
+)
+
+// DefaultFastFail is the default value for fast fail.
+const DefaultFastFail = false
+
+var (
+	// DefaultMaxConcurrentCmds is the default value for the maximum
+	// number of concurrent commands.
+	DefaultMaxConcurrentCmds = runtime.NumCPU()
+	// DefaultEventHandler is the default Event handler.
+	DefaultEventHandler = logEvent
+	// DefaultClock is the default function to use as a clock.
+	DefaultClock = time.Now
+)
+
+// Event is an event that happens during the runner's Run call.
+type Event struct {
+	Type   EventType              `json:"type,omitempty" yaml:"type,omitempty"`
+	Time   time.Time              `json:"time,omitempty" yaml:"time,omitempty"`
+	Fields map[string]interface{} `json:"fields,omitempty" yaml:"fields,omitempty"`
+	Error  string                 `json:"error,omitempty" yaml:"error,omitempty"`
+}
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runner)
+
+// WithFastFail returns a RunnerOption that will return error fun
+// Run as soon as one of the commands fails.
+func WithFastFail() RunnerOption {
```

Comment:
Hmm not sure. I kind of want to leave this as is just because it looks cleaner and more consistent, I know that's a silly reason but.